ENH: Adds new action annotate

VinzentRisch opened the pull request:
closes #3 
depends on [#345](https://github.com/qiime2/q2-types/pull/345)

- Adding new action that can annotate `SampleData[MAGs]` `FeatureData[MAG]` and `SampleData[Contigs]` with AMR Genes.
- The sequence input can be combined with options to input `GenomeData[Proteins]` and `GenomeData[Loci]`.
- The action can also be run with `GenomeData[Proteins]` alone.
- The `GenomeData[Proteins]` and `GenomeData[Loci]` can be created with the action predict-prodigal of q2-moshpit.
- The outputs are `GenomeData[AMRFinderplusAnnotation]` once for annotations and once for mutations.
- Additionally `GenomeData[Proteins]` and Genes for the predicted AMR genes and proteins. These outputs are empty if the sequences or the proteins input is not given.

[Documentation for Amrfinderplus](https://github.com/ncbi/amr/wiki/Running-AMRFinderPlus)

### Set up an environment 
```
mamba create -y -n q2-amrfinderplus -c https://packages.qiime2.org/qiime2/2024.10/metagenome/passed/ -c conda-forge -c bioconda --strict-channel-priority qiime2 q2-types q2-feature-table q2cli ncbi-amrfinderplus
```


### Run it locally 

1. First, clone the repo and checkout the PR branch:
```bash
git clone https://github.com/bokulich-lab/q2-amrfinderplus.git
cd q2-amrfinderplus
git fetch origin pull/4/head:pr-4
git checkout pr-4
pip install -e .
```
```bash
git clone https://github.com/qiime2/q2-types.git
cd q2-types
git fetch origin pull/345/head:pr-345
git checkout pr-345
pip install -e .
```

2. Test data
[PR_4.zip](https://github.com/user-attachments/files/17128253/PR_4.zip)



3. Test it out!
```bash
# download db
qiime amrfinderplus fetch-amrfinderplus-db --o-amrfinderplus-db amrfinderplus_db.qza

# sample data mags
qiime amrfinderplus annotate --i-sequences sample_data_mags/mags.qza --i-proteins sample_data_mags/proteins.qza --i-loci sample_data_mags/loci.qza --p-organism "Escherichia" --output-dir sample_data_mags/out_combined --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-proteins sample_data_mags/proteins.qza  --p-organism "Escherichia" --output-dir sample_data_mags/out_proteins --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-sequences sample_data_mags/mags.qza  --p-organism "Escherichia" --output-dir sample_data_mags/out_sequences --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-sequences sample_data_mags/mags.qza --output-dir sample_data_mags/out_without_organism --i-amrfinderplus-db amrfinderplus_db.qza --verbose

# feature data mags
qiime amrfinderplus annotate --i-sequences feature_data_mags/mags.qza --i-proteins feature_data_mags/proteins.qza --i-loci feature_data_mags/loci.qza --p-organism "Escherichia" --output-dir feature_data_mags/out_combined --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-proteins feature_data_mags/proteins.qza  --p-organism "Escherichia" --output-dir feature_data_mags/out_proteins --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-sequences feature_data_mags/mags.qza  --p-organism "Escherichia" --output-dir feature_data_mags/out_sequences --i-amrfinderplus-db amrfinderplus_db.qza --verbose
qiime amrfinderplus annotate --i-sequences feature_data_mags/mags.qza --output-dir feature_data_mags/out_without_organism --i-amrfinderplus-db amrfinderplus_db.qza --verbose

# contigs
qiime amrfinderplus annotate --i-sequences contigs/contigs.qza  --p-organism "Escherichia" --output-dir contigs/out_contigs --i-amrfinderplus-db amrfinderplus_db.qza --verbose
```

ChristosMatzoros reviewed (changes requested):
Hey @VinzentRisch, I executed all the actions and everything seems to work fine. I have no further comments for the code, just a minor change in the description. Thank you for adding the warning messages for the empty artifacts.

VinzentRisch commented:
Hi Michal could you have a look at this this week or at the beginning of next week. 
The release of qiime 2024 is on the 21. and these changes have to be merged one or two weeks before that.

misialq reviewed (changes requested):
Hey @VinzentRisch, here goes the first round! I tried it with some other param combinations and I got one error when trying:
```
qiime amrfinderplus annotate \
  --i-sequences sample_data_mags/mags.qza \
  --i-proteins sample_data_mags/proteins.qza \
  --i-loci sample_data_mags/loci.qza \
  --p-organism "Serratia" \
  --p-annotation-format genbank \
  --output-dir sample_data_mags/out_combined3 \
  --i-amrfinderplus-db amrfinderplus_db.qza \
  --verbose
``` 
It's the `genbank` option which broken it with the following error:
```
Software directory: '/Users/mziemski/miniconda3/envs/q2-amrfinderplus/bin/'
Software version: 3.12.8
Database directory: '/private/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/8aa5c26a-8b6c-4f46-ac8a-91a3fbff09ee/data'
Database version: 2024-07-22.1
AMRFinder combined translated and protein and mutation search

GFF file mismatch.
*** ERROR ***
gff_check.cpp: Protein FASTA id "contig01_1" is not in the GFF file

HOSTNAME: ?
SHELL: /bin/zsh
PWD: /Users/mziemski/Downloads/PR_4
PATH: /Users/mziemski/miniconda3/envs/q2-amrfinderplus/bin:/Users/mziemski/miniconda3/condabin:/Users/mziemski/.pyenv/shims:/Users/mziemski/.local/bin:/Users/mziemski/homebrew/bin:/Users/mziemski/_tools/edirect:/Users/mziemski/_tools/seqtk:/Users/mziemski/_tools:/Users/mziemski/.gem/ruby/2.6.0/bin:/Users/mziemski/homebrew/opt/libpq/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/usr/local/MacGPG2/bin:/Applications/Privileges.app/Contents/Resources:/Library/TeX/texbin:/usr/local/munki:/Users/mziemski/.cargo/bin:/Users/mziemski/Library/Application Support/JetBrains/Toolbox/scripts
Progam name:  gff_check
Command line: /Users/mziemski/miniconda3/envs/q2-amrfinderplus/bin/gff_check /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/9ea72b46-a5f3-46c0-948e-00317693bbcf/data/sample1/mag1.gff -gfftype genbank -prot /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/5b563ef2-0534-4bbc-b672-8be40816dfe2/data/sample1/mag1.fasta -dna /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/4977ecb5-22d3-4395-bd1d-0bdf6ca5be32/data/sample1/mag1.fa -log /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T//amrfinder.JPY7yW/log: No such file or directory
Traceback (most recent call last):
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2_amrfinderplus/utils.py", line 147, in _run_amrfinderplus_analyse
    run_command(cmd=cmd)
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2_amrfinderplus/utils.py", line 22, in run_command
    subprocess.run(cmd, check=True, cwd=cwd)
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['amrfinder', '--database', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/8aa5c26a-8b6c-4f46-ac8a-91a3fbff09ee/data', '-o', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-AMRFinderPlusAnnotationsDirFmt-5bh46z2o/sample1/mag1_amr_annotations.tsv', '--print_node', '-n', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/4977ecb5-22d3-4395-bd1d-0bdf6ca5be32/data/sample1/mag1.fa', '--nucleotide_output', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-GenesDirectoryFormat-bp2v1z6e/sample1/mag1_amr_genes.fasta', '-p', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/5b563ef2-0534-4bbc-b672-8be40816dfe2/data/sample1/mag1.fasta', '--protein_output', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-ProteinsDirectoryFormat-6evx_1kt/sample1/mag1_amr_proteins.fasta', '-g', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/qiime2/mziemski/data/9ea72b46-a5f3-46c0-948e-00317693bbcf/data/sample1/mag1.gff', '--organism', 'Serratia', '--mutation_all', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-AMRFinderPlusAnnotationsDirFmt-yx2460ga/sample1/mag1_amr_all_mutations.tsv', '--coverage_min', '0.5', '--translation_table', '11', '--annotation_format', 'genbank', '--gpipe_org']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2cli/commands.py", line 530, in __call__
    results = self._execute_action(
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2cli/commands.py", line 596, in _execute_action
    results = action(**arguments)
  File "<decorator-gen-4>", line 2, in annotate
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/qiime2/sdk/action.py", line 289, in bound_callable
    outputs = self._callable_executor_(
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/qiime2/sdk/action.py", line 568, in _callable_executor_
    output_views = self._callable(**view_args)
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2_amrfinderplus/annotate.py", line 113, in annotate
    _run_amrfinderplus_analyse(
  File "/Users/mziemski/miniconda3/envs/q2-amrfinderplus/lib/python3.10/site-packages/q2_amrfinderplus/utils.py", line 149, in _run_amrfinderplus_analyse
    raise Exception(
Exception: An error was encountered while running AMRFinderPlus, (return code 1), please inspect stdout and stderr to learn more.

Plugin error from amrfinderplus:

  An error was encountered while running AMRFinderPlus, (return code 1), please inspect stdout and stderr to learn more.

See above for debug info.
```
Have you seen that before?

VinzentRisch commented:
The error with --p-annotation-format genbank is to be expected. 
The --p-annotation-format is to specify the format of the gff file that is given as input. I guess you used the test data that i provided? That GFF file has a prodigal format and with --p-annotation-format genbank  it throws an error I think.

VinzentRisch commented on q2_amrfinderplus/annotate.py:
```diff
@@ -0,0 +1,129 @@
+    ident_min: float = None,
+    curated_ident: bool = False,
+    coverage_min: float = 0.5,
+    translation_table: str = "11",
```
Yes it is always an integer. But with integers you can only use the function Range() where you can specify a range of integers with start and end. But because there are numbers missing from 17 to 20 a range from 1 to 33 is not appropriate. A workaround is to specify it as strings and with that you can use Choices(). This does not change anything for the user as in the cli you can still specify translation_table without "" and it still works.

misialq replied:
Alright, makes sense!

VinzentRisch commented on q2_amrfinderplus/annotate.py:
```diff
@@ -0,0 +1,129 @@
+    common_params = locals().copy()
+    del common_params["sequences"]
+    del common_params["proteins"]
+    del common_params["loci"]
```
Yes I already implemented that just before because I saw it in moshpit. You are right looks a bit better.

misialq commented:
I see, ok, thanks for the explanation! But then, if you know that this should not happen together, I think you should validate that too. It's really confusing for the user to get this error message which does not really say much - assume, I have not read the AMRFinderPlus docs 😆

VinzentRisch commented on q2_amrfinderplus/utils.py:
```diff
@@ -15,3 +20,269 @@ def run_command(cmd, cwd=None, verbose=True):
+    else:
+        proteins.pathspec = r".+\.(fa|faa|fasta)$"
+
+        # Monkey patch the sample_dict instance method of MultiMAGSequencesDirFmt to
+        # ProteinsDirectoryFormat if it has a sample data dir structure
+        if any(item.is_dir() for item in proteins.path.iterdir()):
+            proteins.sample_dict = MultiMAGSequencesDirFmt.sample_dict.__get__(
+                proteins, ProteinsDirectoryFormat
+            )
+            sample_dict = proteins.sample_dict()
+        # Monkey patch the feature_dict instance method of MAGSequencesDirFmt to
+        # ProteinsDirectoryFormat if it has a feature data dir structure
+        else:
+            proteins.feature_dict = MAGSequencesDirFmt.feature_dict.__get__(
+                proteins, ProteinsDirectoryFormat
+            )
+            file_dict = proteins.feature_dict()
+            # create sample_dict with fake sample
+            sample_dict = {"": file_dict}
```
@misialq I added it here can you take a look?
https://github.com/qiime2/q2-types/pull/345

VinzentRisch commented on q2_amrfinderplus/utils.py:
```diff
@@ -15,3 +20,269 @@ def run_command(cmd, cwd=None, verbose=True):
+
+        # If proteins are provided, construct the expected protein file path.
+        if proteins:
+            protein_path = proteins.path / sample_id / f"{id}.fasta"
```
What do you mean with safe? 
- Most python packages accept pathlib objects the only thing to remember is it has to be a string when a command for subprocess.run is created.
- Pathlib offers much better path manipulation than os.path.
- If you do .path on a dir format it directly gives you a Path object.
- And it makes everything shorter and more readable.

misialq replied:
My comment is not about pathlib objects - it's about how you construct your proteins_path altogether. Instead of manually inserting slashes, you can just do: `protein_path = os.path.join(proteins.path, sample_id, f"{id}.fasta")` - it's as readable as what you did (and not all that much longer) but has additional benefits of:
- being explicit about a path being constructed here (as indicated by the method name)
- being cross-platform compatible (keep in mind that paths are not constructed in the same way on all OS's)
- making sure that any double-slashes are corrected (for example, if one of the elements of the path had an additional slash at the end already)

VinzentRisch commented:
Hi @misialq I implemented all your changes.
For the error when choosing the wrong gff format that you encountered. I cant validate if the gff file and the chosen format match. I added a new exception that is displayed when amrfinerplus prints a "GFF mismatch error" with some additional info.
Could you have another look today or tomorrow. I'm sorry for the time pressure but this should be done as fast as possible so Liz has enough time to add it to the distro before the new release.

misialq commented on q2_amrfinderplus/tests/test_utils.py:
```diff
@@ -41,7 +61,466 @@ def test_run_command_non_verbose(self, mock_print, mock_subprocess_run):
+        )
+
+    def test_create_empty_files_all_true(self):
+        _create_empty_files(
+            sequences=True,
+            proteins=True,
+            organism=True,
+            amr_genes=GenesDirectoryFormat(),
+            amr_proteins=ProteinsDirectoryFormat(),
+            amr_all_mutations=AMRFinderPlusAnnotationsDirFmt(),
+        )
```
We need some kind of an assertion here :)

misialq reviewed (approved):
Approved 🔥 I guess, though, we should wait with merging until the tests pass once the other PR is merged.
